Configurable submodule parameters

ppaulweber opened the pull request:
The built-in `git` resource type does handle repository git submodules with a fixed (static) `--recursive` default behavior which is not always desired in certain projects. The proposed PR introduces a new `in` configuration parameter named `submodule_parameters` to overwrite the default `--recursive` command line argument during `git` `submodule` check-out.

Example usage to check-out all sub-modules but follow the defined `.gitmodule` remote branch without checking out every submodule recursively.

~~~{.yml}
    - get: {{repo}}
      params: 
        submodules: all
        submodule_parameters: --remote
~~~

Example usage to check-out all sub-modules and follow the tracked `git` version state without checking out every submodule recursively.

~~~{.yml}
    - get: {{repo}}
      params: 
        submodules: all
        submodule_parameters: ""
~~~

This proposed changed was successfully build and deployed at https://hub.docker.com/r/ppaulweber/concourse-git-resource/builds/bnflhzbuyeqd6pmjkgl9a9m/
and tested in an example pipeline with source configuration:

~~~{.yml}
resource_types:
- name: git-resource
  type: docker-image
  source:
    repository: ppaulweber/concourse-git-resource
    tag: feature_submodule_parameters

- name: {{repo}}
  type: git-resource
  source:
    branch: master
    uri: {{repo-uri}}
    private_key: {{repo-auth}}
~~~

pivotal-issuemaster commented:
@ppaulweber Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/git-resource&pullRequestId=186)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/git-resource&pullRequestId=186) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@ppaulweber Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/git-resource&pullRequestId=186)!

vito commented:
Thanks for taking the time to submit this!

Having a pass-through param is very risky though in terms of API design and maintenance - see the ["kitchen sink config"](https://github.com/concourse/concourse/wiki/Anti-Patterns#kitchen-sink-config) antipattern.

We'd likely accept a PR that had higher-level params instead.

We'll also need tests written for any behavioral change/new feature. Here's how to run them: https://github.com/concourse/git-resource#running-the-tests

ppaulweber commented:
@vito thank you for your fast reply and comments

- I've updated the change to support two new submodule options instead -- `submodule_recursive` and `submodule_remote`
- added proper tests
- updated the readme file
- build, testet, and deployed at: 
  https://hub.docker.com/r/ppaulweber/concourse-git-resource/builds/bmaljrcro3tyljeksfp8nbi/

vito commented:
Perfect, thanks a bunch!
